From d02227106088f9399aa56b60e091e9209bc99bae Mon Sep 17 00:00:00 2001
From: Archer <archer@beezig.eu>
Date: Thu, 9 Sep 2021 21:59:52 +0200
Subject: [PATCH] accurate-collisions: Fix water

This commit fixes a bug that'd cause you to survive falling on a blocks
edge if part of your hitbox is above water.

diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 0c2b7d5bb..70d94a9ec 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -959,7 +959,29 @@ public abstract class Entity implements ICommandListener {
         // KigPaper start
         if (PaperSpigotConfig.accurateBlockCollisions) {
             // Account for vertical speed in minY
-            boundingBox = new AxisAlignedBB(boundingBox.a, boundingBox.b + this.motY, boundingBox.c, boundingBox.d, boundingBox.e, boundingBox.f);
+            AxisAlignedBB tmp = new AxisAlignedBB(boundingBox.a, boundingBox.b + this.motY / 2, boundingBox.c, boundingBox.d, boundingBox.e, boundingBox.f);
+
+            boolean useTmp = true;
+            int tmpX = MathHelper.floor(tmp.a);
+            int tmpMaxX = MathHelper.floor(tmp.d);
+            int tmpY = MathHelper.floor(tmp.b);
+            int tmpMaxY = MathHelper.floor(tmp.e);
+            int tmpZ = MathHelper.floor(tmp.c);
+            int tmpMaxZ = MathHelper.floor(tmp.f);
+            BlockPosition.MutableBlockPosition blockPosition = new BlockPosition.MutableBlockPosition();
+            Block block;
+            for (int x = tmpX; x <= tmpMaxX; ++x) {
+                for (int y = tmpY; y <= tmpMaxY; ++y) {
+                    for (int z = tmpZ; z <= tmpMaxZ; ++z) {
+                        block = world.getType(blockPosition.c(x, y, z)).getBlock();
+
+                        if (block.getMaterial().isSolid()) {
+                             useTmp = false;
+                         }
+                    }
+                }
+            }
+            if (useTmp) boundingBox = tmp;
         }
         // KigPaper end
         if (this.world.a(boundingBox.grow(0.0D, -0.4000000059604645D, 0.0D).shrink(0.001D, 0.001D, 0.001D), Material.WATER, this)) {
-- 
2.33.0

